### ISSUE TITLE
backend/ftp: missing full modified time support

### ISSUE THREAD

**Issue by ivandeex** [rclone/rclone#5283]
#### The associated forum post URL from `https://forum.rclone.org`

N/A

#### What is your current rclone version (output from `rclone version`)?

rclone v1.55.1
- os/version: ubuntu 18.04 (64 bit)

#### What problem are you are trying to solve?

SetModTime in the FTP backend is a stub, the touch command does nothing, etc.
This also prevents from testing new bisync with this backend due to the way bisync tests work, namely by comparing literal time on files under test with golden results.

#### How do you think rclone should be changed to solve that?

The core issue is with underlying jlaffaye/ftp library, which does not support the [MFMT](https://www.serv-u.com/resource/tutorial/dsiz-mfct-mfmt-avbl-pass-xpwd-xmkd-ftp-command) extension present in most modern FTP servers like `vsftpd`, `FileZilla` etc.

I prepared a [patch for jlaffaye/ftp](https://github.com/ivandeex/ftp/commits/master) to submit as a pull request as well as relevant [rclone patch](https://github.com/ivandeex/rclone/commits/ci-ftpmfmt).

Additionally I found that `rclone serve ftp` also lacks `MFMT` feature. Moreover, it lacks `MLSD` extension, which allows to query the full ISO file time. The legacy `LIST` command, which is the only listing command supported currently, is very limited w.r.t file times. It can return either time with 1 second precision but without year for recently modified files, or date with year but without daytime. I prepared a quick fix for the [goftp.io/server](https://github.com/ivandeex/goftp-server/commits/master) library, however this might have been fixes in its v2.

The fixes will be submitted in due course.

#### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

**Comment by ncw** [comment]
> SetModTime in the FTP backend is a stub, the touch command does nothing, etc.
> This also prevents from testing new bisync with this backend due to the way bisync tests work, namely by comparing literal time on files under test with golden results (#5164).

SetModTime will prevent modtime syncing also so adding it is very useful.
 
> #### How do you think rclone should be changed to solve that?
> The core issue is with underlying `jlaffaye/ftp` library, which does not support the [MFMT](https://www.serv-u.com/resource/tutorial/dsiz-mfct-mfmt-avbl-pass-xpwd-xmkd-ftp-command) extension present in most modern FTP servers like `vsftpd`, `FileZilla` etc.
> 
> I prepared a [patch for jlaffaye/ftp](https://github.com/ivandeex/ftp/commits/master) to submit as a pull request as well as

That patch looks excellent to me :+1:

> relevant [rclone patch](https://github.com/ivandeex/rclone/commits/ci-ftpmfmt).

And so does that. Do the integration tests pass with it against an FTP server which supports MFMT?
 
> Additionally I found that `rclone serve ftp` based on `goftp.io/server` also lacks the `MFMT` feature. Moreover, it lacks the `MLSD` extension, which allows to query the full ISO file time. The legacy `LIST` command, which is the only listing flavor supported currently, is very limited w.r.t file times. It can return either time with 1 second precision but without year (sic!) for recently modified files, or a date with year but without daytime! I prepared a [quick fix for goftp.io/server](https://github.com/ivandeex/goftp-server/commits/master), however this might have been already fixed in its `v2`.

Your patch looks good, but it does break backwards compatibility...

Also there is a `v2`! We could do with a tool which tells us when there is a new major version of dependencies since I don't think we notice otherwise.

I saw a MLSD patch in the v2 branch.
 
> The fixes will be submitted in due course.

:-)

This looks really useful - thanks :-)

**Comment by ivandeex** [comment]
> And so does that. Do the integration tests pass with it against an FTP server which supports MFMT?

Yes. FileZilla in three modes (plain FTP, implicit or explicit TLS) and FTP-something embedded in my WiFi station.

> Your patch looks good, but it does break backwards compatibility. Also there is a v2!

I envisage the goftp-server maintainers will bring me to reason as the play progresses.
